src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -988,28 +1021,76 @@ private async Task SendWithLock(ConnectionState expectedConnectionState, HubMess
         return null;
     }
 
-    private async Task DispatchInvocationAsync(InvocationMessage invocation)
+    private async Task DispatchInvocationAsync(InvocationMessage invocation, ConnectionState connectionState)
     {
+        var expectsResult = !string.IsNullOrEmpty(invocation.InvocationId);
         // Find the handler
         if (!_handlers.TryGetValue(invocation.Target, out var invocationHandlerList))
         {
             Log.MissingHandler(_logger, invocation.Target);
+            if (expectsResult)
+            {
+                Log.NoResultGiven(_logger, invocation.Target, invocation.InvocationId!);
+                await SendWithLock(connectionState, CompletionMessage.WithError(invocation.InvocationId!, "Client didn't provide a result."), cancellationToken: default).ConfigureAwait(false);
+            }
             return;
         }
 
         // Grabbing the current handlers
         var copiedHandlers = invocationHandlerList.GetHandlers();
+        object? result = null;
+        Exception? resultException = null;
+        var hasResult = false;
         foreach (var handler in copiedHandlers)
         {
             try
             {
-                await handler.InvokeAsync(invocation.Arguments).ConfigureAwait(false);
+                var task = handler.InvokeAsync(invocation.Arguments);
+                if (handler.HasResult && task is Task<object?> resultTask)
+                {
+                    if (hasResult)
+                    {
+                        Log.IgnoringPreviousResult(_logger, invocation.Target);
```

Comment:
I would rather we throw preemptively when the second `On` callback with a return value is added for a given method.